### ISSUE TITLE
WebRTC meeting record

### ISSUE THREAD

**Issue by sumboid** [hardcoreeng/platform#153]
Kurento by itself can record endpoints and grid hub that could combine multiple endpoint streams. Hub is not configurable, and it is not enough to have just grid for us.

Just to not forget the idea. We can manage kurento to record all streams separately and then combine them to make single video.
To post process, we can use ffmpeg set of libraries, most of the work would be related to using filters, and in case we want to make it in TS, we can use https://github.com/Streampunk/beamcoder bindings.
Here are multiple filters that we can use for our case (full documentation is here: https://ffmpeg.org/ffmpeg-filters.html):
1. [drawtext](https://ffmpeg.org/ffmpeg-filters.html#drawtext-1) to render text
2. [overlay](https://ffmpeg.org/ffmpeg-filters.html#overlay-1) to overlay text, image or anything we want
3. [xstack](https://ffmpeg.org/ffmpeg-filters.html#xstack) to stack videos
4. [crop](https://ffmpeg.org/ffmpeg-filters.html#crop), [scale](https://ffmpeg.org/ffmpeg-filters.html#scale-1) to fit videos
5. [concat](https://ffmpeg.org/ffmpeg-filters.html#concat) to concatenate video parts of specific tile
6. [amix](https://ffmpeg.org/ffmpeg-filters.html#amix) to mix audio of all streams
7. [acompressor](https://ffmpeg.org/ffmpeg-filters.html#acompressor), [alimiter](https://ffmpeg.org/ffmpeg-filters.html#alimiter) are optional. To make mixed audio to not clip that would lead to distortion. It is optional because usually only one person is talking at the same time.

As for architecture of service, we still need an example to understand how much time it take to postprocess meeting, how it depends on participants and duration, but basically service should have a queue, and as input it could take metadata (timestamps: [start, end], filename, kind: cam | screen, participant info: {name, isOwner}), and later all streams can be uploaded. After service has everything to process video it could add it to the processing queue.

**Comment by haiodo** [comment]
One question here, is it possible to do recording on customer side? 

So If I'm on meeting, I've click record and have video file produced on my local file system?

**Comment by sumboid** [comment]
@haiodo yes, in theory we can try to use [Recording API](https://developer.mozilla.org/en-US/docs/Web/API/MediaStream_Recording_API) that can take MediaStream and record it.
But there are multiple limitations:
1. MediaStream that we can use could be screen/window/tab, specific video from meeting and AFAIK canvas. In case I do this I'd prefer to record specific DOM element, but even this will be supported we likely force user to keep meeting opened (in Platform tab or different tab).
2. AFAIK audio record is not fully supported in screen case, at least I haven't found such feature in screen recorder chromium plugins

Maybe we can research how we can post-process set of video streams on client side, but likely there are not many ways to do this. Also: https://github.com/Kagami/ffmpeg.js and https://github.com/ffmpegwasm/ffmpeg.wasm

**Comment by sumboid** [comment]
To not forget, there is book with nice ffmpeg examples: http://www.astro-electronic.de/FFmpeg_Book.pdf
